docs/architecture/adr-083-libp2p-implementation.md

```diff
@@ -0,0 +1,163 @@
+branch--[`main/libp2p`](https://github.com/tendermint/tendermint/tree/main/libp2p)--to
+prototype the new libp2p integration. This work will aim to:
+
+- develop a working implementation (by way of the e2e tests) of the
+  p2p layer without requiring changes to the reactor implementations.
+- migrate the reactor tests after completing a prototype. 
+- address design questions as they come up, by amending this document
+  or in periodic conversations over patches. 
+- minimize or avoid any interface changes between the reactor
+  implementations and the peer layer. 
+
+### Use libp2p Idiomatically 
+
+We should endeavor to use Libp2p as idiomatically as possible and take
+advantage of features and developments available in the libp2p
+ecosystem.
+
+### Minimize P2P Code
+
+One of the largest benefits of this project is that we should be able
+to delete the vast majority of the existing p2p code, which should
+reduce the maintenance burden within Tendermint.
+
+### Defer Transport Decisions
+
+There are many high level aspects of p2p systems such as handshakes
+between nodes, transport protocols, which are fully configurable in
+the context of libp2p. There is no need to make decisions about these
+at this time, as these aspects of the system should be below the level
+of concern for tendermint.
+
+We will not make these decisions user-configurable without a specific
+need.  We will make it possible for users to pass a custom libp2p
+connection object (`host.Host`) with limited support, but will
+otherwise not expose configurations and customizations of different
+libp2p fundamentals.
+
+### Use GossipSub for Gossip Communication 
+
+An unfortunate consequence of the current p2p implementation is that
+reactors (e.g., mempool, consensus, and evidence) implement their own
+gossip logic, and maintain their own tracking of the connected peer
+network. These workloads should eventually migrate to using gosipsub
+(e.g., setting `broadcast: true` and letting libp2p broadcast,)
+however, this requires removing all of the legacy code. In the short
+term, we should leave the legacy gossip implementation for the first
+phase of the project. 
+
+## Detailed Design
+
+Based on what we currently know about this process, we will:
+
+- complete basic implementation of P2P fundamentals (e.g. Channel,
+  host connection, etc.) for libp2p with the goal of having e2e
+  support.
+  
+- add support for running libp2p e2e test networks, and add this as a
+  branch in the test configuration and enable nightly testing of the
+  libp2p branch. 
+
+- migrate all unit tests away from using the legacy router, and engage
+  in some design work on some kind of new test harness. Consider
+  reviving a shimed routing layer or using libp2p's features, for
+  testing purposes. It might also be possible to actually use real
+  networks for testing.
+
+- delete all legacy p2p code. 
+
+- migrate all gossip workloads to use gossipsub. 
```

Comment:
A lot of us have been over our current implementation and reviewed the gossipsub implementation and feel pretty confident about it. If it turns out to be wrong, libp2p offers different pubsub mechanisms, so we have options (and swapping this out would be rather trivial.) 

Our legacy system, generally lacks any guarantees, and while I think it's possible to look at it and see something that appears to be designed, in practice it's performance/capability is a side effect of its implementation (e.g. subtle changes to the networking configuration and buffering patterns will cause nodes to fall off the network.) so I don't think we have a lot of ground to stand on here.

Reply:
All the files in https://github.com/libp2p/specs/tree/master/pubsub/gossipsub. PubSub and all its extensions are speced out, so you can explore all the details and guarantees it provides.

Reply:
I superficially read the specification for GossipSub, it really recalls (without mentioning, from what I noticed) the classic gossip protocols introduced by [Bimodal Multicast](https://dl.acm.org/doi/abs/10.1145/312203.312207). I don't see how this technical spec can provide any guarantees, which in gossip are essentially probabilistic, but it provides a high-level view of the dissemination protocol.

Reply:
I reopened this conversation because, while I think we should try to migrate as much functionalities as possible to existing gossip propagation mechanisms, we will not able to migrate *all* workloads to any generic solution without an important and costly redesign of the protocols that generate them, and consensus is only the most obvious example of this.

Reply:
In summary, from what I see from this discussion and for the sake of having this ADR approved, we could just not promise to do what we cannot ensure we will be able to do. In other words, we can try to migrate *some* workloads to gossipsub, not compromise to migrate *all* workloads to it.